client/i18n/language-detector.ts

```diff
@@ -0,0 +1,31 @@
+import { ALL_TRANSLATION_LANGUAGES, TranslationLanguage } from '../../common/i18n'
+
+const lowerCaseSupportedLanguages = ALL_TRANSLATION_LANGUAGES.map(l => l.toLowerCase())
+
+/**
+ * Choose the preferred language from the list of given languages. If no languages are given, we
+ * default to using `navigator.languages` as a list of available languages. The logic for choosing
+ * the preferred language is:
+ *  - for languages we support without dialects, prefer the earliest matching language in the list
+ *    of given languages.
+ *  - for languages we support with dialects, prefer only if exact match.
+ */
+export function getBestLanguage(languages = navigator.languages) {
```

Comment:
Not making the available languages a parameter means that the unit tests you've written a brittle to changes with that (and it's hard to test cases that may not exist yet just because we don't support some particular language), so I would pass that in as a parameter as well (and update your tests to use some constant list)